dll/win32/msafd/misc/dllmain.c

```diff
@@ -1962,58 +2020,97 @@ WSPConnect(SOCKET Handle,
+                               NULL,
+                               SynchronizationEvent,
+                               FALSE);
+        if (!NT_SUCCESS(Status))
+            goto Leave;
+
+        APCContext = HeapAlloc(GlobalHeap, 0, sizeof(AFDCONNECTAPCCONTEXT));
+        if (!APCContext)
+        {
+            ERR("Not enough memory for APC Context\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Leave;
+        }
+        APCContext->lpConnectInfo = ConnectInfo;
+        APCContext->lpSocket = Socket;
+        APCContext->hSockEvent = SockEvent2;
+        APCFunction = &AfdConnectAPC; // should be a private io completition function inside us
+
+        IOSB = &APCContext->IoStatusBlock;
     }
 
+    IOSB->Status = STATUS_PENDING;
+
     /* Send IOCTL */
     Status = NtDeviceIoControlFile((HANDLE)Handle,
-                                   SockEvent,
-                                   NULL,
-                                   NULL,
-                                   &IOSB,
+                                   APCFunction != NULL ? NULL : SockEvent,
+                                   APCFunction,
+                                   APCContext,
+                                   IOSB,
                                    IOCTL_AFD_CONNECT,
                                    ConnectInfo,
                                    0x22,
                                    NULL,
                                    0);
-    /* Wait for return */
     if (Status == STATUS_PENDING)
     {
-        WaitForSingleObject(SockEvent, INFINITE);
-        Status = IOSB.Status;
+        if (Socket->SharedData->NonBlocking)
+        {
+            Status = IOSB->Status;
+            if (Status == STATUS_PENDING)
+            {
+                Socket->SharedData->State = SocketConnected;
+                Status = STATUS_CANT_WAIT; // WSAEWOULDBLOCK
+                goto Leave;
+            }
+        }
+        else
+        {
+            /* Wait for completion of blocking mode */
+            WaitForSingleObject(SockEvent, INFINITE);
+            Status = IOSB->Status;
+        }
     }
 
-    Socket->SharedData->SocketLastError = TranslateNtStatusError(Status);
-    if (Status != STATUS_SUCCESS)
-        goto notify;
+    if (!Socket->SharedData->NonBlocking)
+    {
+        Socket->SharedData->SocketLastError = TranslateNtStatusError(Status);
+        if (Status != STATUS_SUCCESS)
+            goto Leave;
 
-    Socket->SharedData->State = SocketConnected;
-    Socket->TdiConnectionHandle = (HANDLE)IOSB.Information;
-    Socket->SharedData->ConnectTime = GetCurrentTimeInSeconds();
+        Socket->SharedData->State = SocketConnected;
+        Socket->TdiConnectionHandle = (HANDLE)IOSB->Information;
+        Socket->SharedData->ConnectTime = GetCurrentTimeInSeconds();
+    }
 
     /* Get any pending connect data */
-    if (lpCalleeData != NULL)
+    if (lpCalleeData != NULL && Status == STATUS_SUCCESS)
```

Comment:
Should these Status checks be really against STATUS_SUCCESS and nothing else? (like for example `NT_SUCCESS(Status)`)

Reply:
Not sure, the check is the same as original.

[https://github.com/reactos/reactos/blob/1ad90967d709687a70d56da5fac12d28f6577b74/dll/win32/msafd/misc/dllmain.c#L2002](https://github.com/reactos/reactos/blob/1ad90967d709687a70d56da5fac12d28f6577b74/dll/win32/msafd/misc/dllmain.c#L2002)

Reply:
I wonder whether you could add a DPRINT1() to detect cases where the condition `NT_SUCCESS(Status) && (Status != STATUS_SUCCESS)` is TRUE, if this could help understanding in which cases this happens.